### PR TITLE
Object Embedding Refactor

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Ports over TG's object embedding system.

This is a refactor/fix of the previous system (which is currently non-existent as of the latest move refactor).

This basically is similar to the previous system, only there's a lot more control over how often something embeds and how much damage it does once it does embed.

The system features a thrown alert for if you have an embedded object in yourself and allows you to yank it out with said alert. Yanking objects out of other people is *not* possible except via surgery.

Yanking it out yourself incurs damage and makes you scream (unlike the current system).

This does impact bullet shrapnel quite a bit (you can dig into your own wound and yank it out), but I'm inclined to say this is probably a fair trade-off, as this system incurs buttloads of brute damage on someone who has an embedded object in them.

Lastly, implements TG's traitor throwing stars; they have a 100% embed chance and cause a lot of damage if they're left inside you. They come in a kit with a few of them and two traitors boalas.

Changes from TG:
 - Currently embedded objects don't cause increased bleeding: this is purely because of the active blood refactor, at the moment; as soon as that is finalized, this will be added, as well.

Fixes/changes

- Fixes the major clusterbutt that is the various implant surgeries
 -  Implants only remove actual implants
 -  Cavity implants are now properly put in the ride holding var as opposed to internal organs (WTF)
 -  Embedded object removal is its own distinctive surgery
 -  Cortical borers are handled as an organ manipulation surgery if done on the head (at the extract organ stage)
 - Removed the vestiges of pinning code
 - Bullets no longer cause shrapnel

:cl: Fox McCloud
fix: Re-enables object embedding system; embedding objects now cause more damage than before
add: Adds in throwing stars kit to the traitor uplink
rscdel: bullets no longer leave shrapnel
/:cl:

### CODE REVIEW

File: code/__DEFINES/combat.dm [hunk 1]
```diff
@@ -63,13 +63,21 @@
...
 //Gun Stuff
  #define SAWN_INTACT  0
  #define SAWN_OFF     1
 
  #define WEAPON_LIGHT 0
  #define WEAPON_MEDIUM 1
- #define WEAPON_HEAVY 2
-
-// Embedded objects
-#define EMBED_THROWSPEED_THRESHOLD 15
```
Comment (Fox-McCloud) [inline code/__DEFINES/combat.dm]: At first I thought something went horrendously wrong with the refactor; then I noticed this.

File: code/modules/mob/living/carbon/human/death.dm [hunk 2]
```diff
@@ -102,21 +104,8 @@
...
-
-	if(istype(head))
-		for(var/I in head.implants)
-			if(istype(I,/mob/living/simple_animal/borer))
-				B = I
-	if(B)
-		if(B.controlling && B.host == src)
-			B.detach()
-
-		verbs -= /mob/living/carbon/proc/release_control
```
Comment (Fox-McCloud) [inline code/modules/mob/living/carbon/human/death.dm]: I believe all the above is no longer needed for borers with the semi-recent refactor.

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 3]
```diff
@@ -46,13 +46,14 @@ emp_act
...
-				(SP.desc) = "[SP.desc] It looks like it is a [P.ammo_casing.caliber] caliber round."
+				SP.desc = "[SP.desc] It looks like it is a [P.ammo_casing.caliber] caliber round."
 			else
-				(SP.desc) = "[SP.desc] The round's caliber is unidentifiable."
-			(SP.loc) = organ
-			organ.embed(SP)
+				SP.desc = "[SP.desc] The round's caliber is unidentifiable."
+			throw_alert("embeddedobject", /obj/screen/alert/embeddedobject)
+			organ.embedded_objects |= SP
+			SP.forceMove(src)
```
Comment (Fox-McCloud) [inline code/modules/mob/living/carbon/human/human_defense.dm]: This may have to go, for both functionality and balance reasons.

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 4]
```diff
@@ -325,24 +311,19 @@ emp_act
...
-				//Thrown sharp objects have some momentum already and have a small chance to embed even if the damage is below the threshold
-
-				if(((sharp && prob(damage / (10 * I.w_class) * 100)) || (damage > embed_threshold && prob(embed_chance))) && (I.no_embed == 0))
-					affecting.embed(I)*/
+			if(can_embed(I))
+				if(prob(I.embed_chance))
+					throw_alert("embeddedobject", /obj/screen/alert/embeddedobject)
+					var/obj/item/organ/external/L = pick(bodyparts)
+					L.embedded_objects |= I
+//					I.add_mob_blood(src)//it embedded itself in you, of course it's bloody!
```
Comment (Fox-McCloud) [inline code/modules/mob/living/carbon/human/human_defense.dm]: Adding in commented out code, I know; this is because of the blood refactor; will uncomment once its merged.

**Comment by KasparoVy** [comment]
This is excitingly brutal.

File: code/modules/mob/living/carbon/human/human.dm [hunk 5]
```diff
@@ -700,6 +699,28 @@
...
+			var/obj/item/I = locate(href_list["embedded_object"]) in L.embedded_objects
+			if(!I || I.loc != src) //no item, no limb, or item is not in limb or in the person anymore
+				return
+			var/time_taken = I.embedded_unsafe_removal_time*I.w_class
+			usr.visible_message("<span class='warning'>[usr] attempts to remove [I] from their [L.name].</span>","<span class='notice'>You attempt to remove [I] from your [L.name]... (It will take [time_taken/10] seconds.)</span>")
+			if(do_after(usr, time_taken, needhand = 1, target = src))
+				if(!I || !L || I.loc != src || !(I in L.embedded_objects))
+					return
+				L.embedded_objects -= I
+				L.receive_damage(I.embedded_unsafe_removal_pain_multiplier*I.w_class)//It hurts to rip it out, get surgery you dingus.
```
Comment (Fox-McCloud) [inline code/modules/mob/living/carbon/human/human.dm]: You nasty little thing, I should have noticed you the first time.

should be `receive_damage`

File: code/modules/mob/living/carbon/human/life.dm [hunk 6]
```diff
@@ -958,6 +955,22 @@
...
+		var/obj/item/organ/external/BP = X
+		for(var/obj/item/I in BP.embedded_objects)
+			if(prob(I.embedded_pain_chance))
+				BP.take_damage(I.w_class*I.embedded_pain_multiplier)
+				to_chat(src, "<span class='userdanger'>[I] embedded in your [BP.name] hurts!</span>")
+
+			if(prob(I.embedded_fall_chance))
+				BP.take_damage(I.w_class*I.embedded_fall_pain_multiplier)
+				BP.embedded_objects -= I
+				I.loc = get_turf(src)
```
Comment (Fox-McCloud) [inline code/modules/mob/living/carbon/human/life.dm]: `I.forceMove(get_turf(src))`

File: code/modules/surgery/cavity_implant.dm [hunk 7]
```diff
@@ -199,198 +192,15 @@
 				affected.wounds += I
 				affected.owner.custom_pain("You feel something rip in your [affected.name]!", 1)
 			user.drop_item()
-			target.internal_organs += tool
+			affected.hidden = tool
```
Comment (Fox-McCloud) [inline code/modules/surgery/cavity_implant.dm]: . . . 

ok then.

File: code/modules/surgery/implant_removal.dm [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		if(istype(user.get_item_by_slot(slot_l_hand), /obj/item/weapon/implantcase))
+			case = user.get_item_by_slot(slot_l_hand)
+		else if(istype(user.get_item_by_slot(slot_r_hand), /obj/item/weapon/implantcase))
+			case = user.get_item_by_slot(slot_r_hand)
+		else
+			case = locate(/obj/item/weapon/implantcase) in get_turf(target)
+
+		if(case && !case.imp)
+			case.imp = I
+			I.loc = case
```
Comment (Fox-McCloud) [inline code/modules/surgery/implant_removal.dm]: `I.forceMove(case)`

**Comment by pinatacolada** [comment]
Can we keep the thing where you can yank out objects out of someone else? It feels a bit silly i can do it myself with my bare hands, but if its someone else they need to do surgery on me

**Comment by Fox-McCloud** [comment]
Need to do some finalized testing on this and damage numbers.

Need to test out shrapnel too to see just how ridiculous it is.

**Comment by Aurorablade** [comment]
We apologise for odd surgery things

File: code/modules/surgery/implant_removal.dm [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+/datum/surgery_step/extract_implant
+	name = "extract implant"
+	allowed_tools = list(/obj/item/weapon/hemostat = 100, /obj/item/weapon/crowbar = 65)
+	time = 64
+	var/obj/item/weapon/implant/I = null
+
+/datum/surgery_step/extract_implant/begin_step(mob/living/user, mob/living/carbon/human/target, target_zone, obj/item/tool,datum/surgery/surgery)
+	var/obj/item/organ/external/affected = target.get_organ(target_zone)
+	I = locate(/obj/item/weapon/implant) in target
```
Comment (pinatacolada) [inline code/modules/surgery/implant_removal.dm]: Dont think you need to do this, end_step is doing it already

File: code/modules/mob/living/carbon/human/human_defense.dm (hunk 3 above)
Comment (pinatacolada) [inline code/modules/mob/living/carbon/human/human_defense.dm]: I dont think we need an alert for embedded bullets, its one of those things that is fun to not know if the bullet got stuck inside or not

File: code/modules/surgery/implant_removal.dm (hunk 9 above)
Comment (Fox-McCloud) [inline code/modules/surgery/implant_removal.dm]: This is just moving around the code; not going to bother fixing any but the most major/glaring of issues.

File: code/modules/mob/living/carbon/human/human_defense.dm (hunk 3 above)
Comment (taukausanake) [inline code/modules/mob/living/carbon/human/human_defense.dm]: Yeah, more than likely you can't pull out a bullet anyways. Anything else should be fine to yank out

**Comment by taukausanake** [comment]
Do you think it would be possible to have an inventory area for embedded items? That would be the best way to pull out things I'm thinking, unless someone else can think something better. I didn't know you could pull out embedded things before

**Comment by Ralta** [comment]
Love the idea, but as suggested above, I'd prefer it for bullet shrapnel to require surgery and to be able to remove embedded objects from someone else. While surgery is obviously preferable, I've had to do it when both ORs are taken and a patient needs stabilising before. Not the biggest deal in the world, but would be nice.

Either way, looking forward to this being merged! 👍

**Comment by SamHPurp** [comment]
Surely the refactor and the traitor items should be in separate PRs.

File: code/__DEFINES/combat.dm (hunk 1 above)
Comment (RemieRichards) [inline code/__DEFINES/combat.dm]: were they actually the same #define name?

File: code/modules/mob/living/carbon/brain/brain_item.dm [hunk 10]
```diff
@@ -71,9 +71,9 @@
 	var/obj/item/organ/internal/brain/B = src
 	if(!special)
 		var/mob/living/simple_animal/borer/borer = owner.has_brain_worms()
-
 		if(borer)
-			borer.detach() //Should remove borer if the brain is removed - RR
+			borer.leave_host() //Should remove borer if the brain is removed - RR
```
Comment (RemieRichards) [inline code/modules/mob/living/carbon/brain/brain_item.dm]: Hi me from 2015(ish)

File: code/datums/spells/summonitem.dm [hunk 11]
```diff
@@ -79,6 +79,14 @@
 								add_logs(target, C, "magically debrained", addition="INTENT: [uppertext(target.a_intent)]")*/
 						if(C.stomach_contents && item_to_retrive in C.stomach_contents)
 							C.stomach_contents -= item_to_retrive
+						for(var/X in C.bodyparts)
+							var/obj/item/organ/external/part = X
+							if(item_to_retrive in part.embedded_objects)
+								part.embedded_objects -= item_to_retrive
+								to_chat(C, "<span class='warning'>The [item_to_retrive] that was embedded in your [part] has myseriously vanished. How fortunate!</span>")
```
Comment (SamHPurp) [inline code/datums/spells/summonitem.dm]: myseriously -> mysteriously

File: code/datums/uplink_item.dm [hunk 12]
```diff
@@ -750,6 +750,14 @@ var/list/uplink_items = list()
 	cost = 17
 	excludefrom = list(/datum/game_mode/nuclear)
 
+/datum/uplink_item/stealthy_weapons/throwingweapons
+	name = "Box of Throwing Weapons"
```
Comment (SamHPurp) [inline code/datums/uplink_item.dm]: This should 100% be in its own PR.

File: code/datums/uplink_item.dm [hunk 13]
```diff
@@ -750,6 +750,14 @@ var/list/uplink_items = list()
...
+	desc = "A box of shurikens and reinforced bolas from ancient Earth martial arts. They are highly effective \
```
Comment (SamHPurp) [inline code/datums/uplink_item.dm]: Not fully up on my lore, but...

Sol, not Earth, no?

File: code/game/objects/items/weapons/weaponry.dm [hunk 14]
```diff
@@ -130,6 +130,21 @@ obj/item/weapon/wirerod/attackby(obj/item/I, mob/user, params)
 		qdel(I)
 		qdel(src)
 
+/obj/item/weapon/throwing_star
+	name = "throwing star"
+	desc = "An ancient weapon still used to this day due to it's ease of lodging itself into victim's body parts"
```
Comment (SamHPurp) [inline code/game/objects/items/weapons/weaponry.dm]: "parts" at the end is superfluous

**Comment by deep-fried-faceplant** [comment]
The throwing stars should be a part of their own PR, as well as any changes that don't directly link to the embedding refactor. While some people might agree with the embedding refactor, others might not agree with the addition of the throwing stars.

Be sure to test rubbershot from shotguns. Rubber pellets embedding in a target can lead to some nasty, unwanted and unintentional results.

**Comment by taukausanake** [comment]
> Rubber pellets embedding

Sounds like working as intended to me

**Comment by deep-fried-faceplant** [comment]
If rubber pellets embed too much, rubber shot becomes much more lethal than intended. Lethal rubber shot was funny for awhile, until it became almost as good as buckshot because it was "non-lethal", stunned, broke bones, and caused embedding.

File: code/game/objects/items/weapons/weaponry.dm (hunk 14 above)
Comment (KasparoVy) [inline code/game/objects/items/weapons/weaponry.dm]: victims' bodies

**Comment by Fox-McCloud** [comment]
Last of the updates; went ahead and removed bullet shrapnel; it's just not suited to this new system, and, even if it was, it generates really nasty feedback loops that are going to get exponentially worse when @Aurorablade's TG port of the blood refactor is merged in (which makes object embedding wayyy worse).

Lore wise, you can explain it as the bullets going "through and through"; TG even has a really neat system that generates blood spatters that spew out from the opposite of a bullet and can spatter on the floor, too; I'll have to implement that some time to re-inforce this point.

**Comment by Aurorablade** [comment]
@Fox-McCloud you gonna do that blood spatter effect after blood refactor?

**Comment by Fox-McCloud** [comment]
@Aurorablade yeah; I'll probably do a quick pass to fix up the rest of the object embedding mechanics with the blood system, make some adjustments to how bloodloss works, turn on the bloodiness of objects for embedded objects, then look into the gunshot blood spatters.

**Comment by tigercat2000** [comment]
LGTM!

**Comment by KasparoVy** [comment]
Thank you so much.
